File: tools/upstream-sync/sync_test.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"
)

func Test_Sync_copies_managed_paths_and_updates_versions(t *testing.T) {
	// Given
	ctx := context.Background()
	root := t.TempDir()
	remote := createRemote(t)
	writeFile(t, filepath.Join(root, "plugins/example/skills/stale.md"), "stale\n")
	writeFile(t, filepath.Join(root, "plugins/example/.codex-plugin/plugin.json"),
		`{"name":"example","version":"2.3.2"}`)
	configPath := filepath.Join(root, "upstreams.json")
	lockPath := filepath.Join(root, "upstreams.lock.json")
	writeFile(t, configPath, fmt.Sprintf(`{
  "upstreams": [{
    "name": "example",
    "repository": %q,
    "ref": "main",
		"copies": [{
		  "source": "content",
		  "destination": "plugins/example/skills",
		  "removeFrontmatter": ["disable-model-invocation"]
		}],
		"version": {
		  "source": "manifest/plugin.json",
		  "appendCommit": true,
		  "targets": ["plugins/example/.codex-plugin/plugin.json"]
		}
  }]
}`, remote))

	// When
	err := syncConfigured(ctx, root, configPath, lockPath)
	// Then
	if err != nil {
		t.Fatalf("syncConfigured() error = %v", err)
	}
	assertFile(t, filepath.Join(root, "plugins/example/skills/SKILL.md"),
		"---\nname: example\n---\nupstream\n")
	assertMissing(t, filepath.Join(root, "plugins/example/skills/stale.md"))
	commit := gitFixtureOutput(t, remote, "rev-parse", "HEAD")
	assertJSONVersion(t, filepath.Join(root, "plugins/example/.codex-plugin/plugin.json"),
		"2.4.0+upstream."+commit[:12])
	assertLockCommit(t, lockPath, "example", commit)
}

func Test_LoadConfig_rejects_destination_outside_root(t *testing.T) {
	// Given
	path := filepath.Join(t.TempDir(), "upstreams.json")
	writeFile(t, path, `{
  "upstreams": [{
    "name": "escape",
    "repository": "https://example.com/repo.git",
    "ref": "main",
    "copies": [{"source": "content", "destination": "../outside"}]
  }]
}`)

	// When
	_, err := loadConfig(path)

	// Then
	if err == nil || !strings.Contains(err.Error(), "destination") {
		t.Fatalf("loadConfig() error = %v, want destination error", err)
	}
}

func Test_BuildVersion_replaces_existing_build_metadata(t *testing.T) {
	// Given
	version := "2.4.0+vendor.previous"
	commit := "0123456789abcdef"

	// When
	got := buildVersion(version, commit)

	// Then
	want := "2.4.0+upstream.0123456789ab"
	if got != want {
		t.Fatalf("buildVersion() = %q, want %q", got, want)
	}
}

func Test_Sync_preserves_workspace_when_any_clone_fails(t *testing.T) {
	// Given
	ctx := context.Background()
	root := t.TempDir()
	remote := createRemote(t)
	destination := filepath.Join(root, "plugins/example/skills/SKILL.md")
	writeFile(t, destination, "local\n")
	configPath := filepath.Join(root, "upstreams.json")
	writeFile(t, configPath, fmt.Sprintf(`{
  "upstreams": [
    {
      "name": "example",
      "repository": %q,
      "ref": "main",
      "copies": [{"source": "content", "destination": "plugins/example/skills"}]
    },
    {
      "name": "broken",
      "repository": %q,
      "ref": "missing-ref",
      "copies": [{"source": "content", "destination": "plugins/broken/skills"}]
    }
  ]
}`, remote, remote))

	// When
	err := syncConfigured(ctx, root, configPath, filepath.Join(root, "upstreams.lock.json"))

	// Then
	if err == nil {
		t.Fatal("syncConfigured() error = nil, want clone failure")
	}
	assertFile(t, destination, "local\n")
}

func createRemote(t *testing.T) string {
	t.Helper()
	dir := t.TempDir()
	gitRun(t, dir, "init", "-b", "main")
	gitRun(t, dir, "config", "user.name", "Upstream Fixture")
	gitRun(t, dir, "config", "user.email", "fixture@example.com")
	writeFile(t, filepath.Join(dir, "content/SKILL.md"),
		"---\nname: example\ndisable-model-invocation: true\n---\nupstream\n")
	writeFile(t, filepath.Join(dir, "manifest/plugin.json"), `{"version":"2.4.0"}`)
	gitRun(t, dir, "add", ".")
	gitRun(t, dir, "commit", "-m", "fixture")
	return dir
}

func gitRun(t *testing.T, dir string, args ...string) {
	t.Helper()
	command := exec.Command("git", args...)
	command.Dir = dir
	if output, err := command.CombinedOutput(); err != nil {
		t.Fatalf("git %s: %v\n%s", strings.Join(args, " "), err, output)
	}
}

func gitFixtureOutput(t *testing.T, dir string, args ...string) string {
	t.Helper()
	command := exec.Command("git", append([]string{"-C", dir}, args...)...)
	output, err := command.Output()
	if err != nil {
		t.Fatalf("git %s: %v", strings.Join(args, " "), err)
	}
	return strings.TrimSpace(string(output))
}

func writeFile(t *testing.T, path, content string) {
	t.Helper()
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		t.Fatalf("mkdir %s: %v", filepath.Dir(path), err)
	}
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		t.Fatalf("write %s: %v", path, err)
	}
}

func assertFile(t *testing.T, path, want string) {
	t.Helper()
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	if string(data) != want {
		t.Fatalf("%s = %q, want %q", path, data, want)
	}
}

func assertMissing(t *testing.T, path string) {
	t.Helper()
	if _, err := os.Stat(path); !os.IsNotExist(err) {
		t.Fatalf("%s still exists", path)
	}
}

func assertJSONVersion(t *testing.T, path, want string) {
	t.Helper()
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	var manifest struct {
		Version string `json:"version"`
	}
	if err := json.Unmarshal(data, &manifest); err != nil {
		t.Fatalf("decode %s: %v", path, err)
	}
	if manifest.Version != want {
		t.Fatalf("version = %q, want %q", manifest.Version, want)
	}
}

func assertLockCommit(t *testing.T, path, name, want string) {
	t.Helper()
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	var lock struct {
		Upstreams map[string]string `json:"upstreams"`
	}
	if err := json.Unmarshal(data, &lock); err != nil {
		t.Fatalf("decode lock: %v", err)
	}
	if lock.Upstreams[name] != want {
		t.Fatalf("lock commit = %q, want %q", lock.Upstreams[name], want)
	}
}

File: tools/upstream-sync/test/sync.test.ts
import assert from 'node:assert/strict';
import { execFile } from 'node:child_process';
import { access, mkdir, mkdtemp, readFile, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { dirname, join } from 'node:path';
import test from 'node:test';
import { promisify } from 'node:util';

import { parseConfig } from '../src/config.ts';
import { syncConfigured } from '../src/sync.ts';

const execFileAsync = promisify(execFile);

test('syncs managed paths and preserves the manifest format', async () => {
  // Given
  const root = await mkdtemp(join(tmpdir(), 'upstream-sync-test-'));
  const remote = await createRemote();
  await write(root, 'plugins/example/skills/stale.md', 'stale\n');
  await write(root, 'plugins/example/.codex-plugin/plugin.json', '{"name":"example","version":"2.3.2"}');
  await write(root, 'upstreams.json', configFor(remote, 'main'));

  // When
  await syncConfigured({ root, configPath: join(root, 'upstreams.json'), lockPath: join(root, 'lock.json') });

  // Then
  assert.equal(
    await readFile(join(root, 'plugins/example/skills/SKILL.md'), 'utf8'),
    '---\nname: example\n---\nupstream\n',
  );
  assert.match(
    await readFile(join(root, 'plugins/example/.codex-plugin/plugin.json'), 'utf8'),
    /"version":"2\.4\.0\+upstream\.[0-9a-f]{12}"/,
  );
  assert.match(await readFile(join(root, 'lock.json'), 'utf8'), /"example": "[0-9a-f]{40}"/);
});

test('propagates a version into a copied manifest and the repository marketplace', async () => {
  // Given
  const root = await mkdtemp(join(tmpdir(), 'upstream-sync-test-'));
  const remote = await createRemote();
  await write(root, 'plugins/example/.codex-plugin/plugin.json', '{"version":"2.3.2","legacy":true}');
  await write(
    root,
    '.claude-plugin/marketplace.json',
    '{"name":"fixtures","plugins":[{"name":"example","version":"2.3.2","source":"./plugins/example"}]}',
  );
  await write(
    root,
    'upstreams.json',
    JSON.stringify({
      upstreams: [
        {
          name: 'example',
          repository: remote,
          ref: 'main',
          copies: [{ source: 'manifest', destination: 'plugins/example/.codex-plugin' }],
          version: {
            source: 'manifest/plugin.json',
            targets: ['plugins/example/.codex-plugin/plugin.json'],
            marketplaceTargets: ['.claude-plugin/marketplace.json'],
            appendCommit: true,
          },
        },
      ],
    }),
  );

  // When
  await syncConfigured({ root, configPath: join(root, 'upstreams.json'), lockPath: join(root, 'lock.json') });

  // Then
  const manifest = await readFile(join(root, 'plugins/example/.codex-plugin/plugin.json'), 'utf8');
  assert.match(manifest, /^\{"version":"2\.4\.0\+upstream\.[0-9a-f]{12}"\}$/);
  assert.match(
    await readFile(join(root, '.claude-plugin/marketplace.json'), 'utf8'),
    /"name": "example",\n {6}"version": "2\.4\.0\+upstream\.[0-9a-f]{12}"/,
  );
});

test('preserves the workspace when preflight fails', async () => {
  // Given
  const root = await mkdtemp(join(tmpdir(), 'upstream-sync-test-'));
  const remote = await createRemote();
  await write(root, 'plugins/example/skills/SKILL.md', 'local\n');
  await write(root, 'upstreams.json', configFor(remote, 'missing-ref'));

  // When
  await assert.rejects(() =>
    syncConfigured({ root, configPath: join(root, 'upstreams.json'), lockPath: join(root, 'lock.json') }),
  );

  // Then
  assert.equal(await readFile(join(root, 'plugins/example/skills/SKILL.md'), 'utf8'), 'local\n');
});

test('rejects a destination outside the workspace root', () => {
  // Given
  const config = JSON.parse(configFor('https://example.test/repository.git', 'main'));
  config.upstreams[0].copies[0].destination = '../outside';

  // When / Then
  assert.throws(() => parseConfig(config), /destination/);
});

test('recovers a journal-less preparation directory before syncing', async () => {
  // Given
  const root = await mkdtemp(join(tmpdir(), 'upstream-sync-test-'));
  const remote = await createRemote();
  await write(root, 'plugins/example/.codex-plugin/plugin.json', '{"name":"example","version":"2.3.2"}');
  await write(root, 'upstreams.json', configFor(remote, 'main'));
  await mkdir(join(root, '.upstream-sync-transaction'));
  const stalePreparation = join(root, '.upstream-sync-transaction.prepare-999999-stale');
  await mkdir(stalePreparation);

  // When / Then
  await assert.doesNotReject(() =>
    syncConfigured({ root, configPath: join(root, 'upstreams.json'), lockPath: join(root, 'lock.json') }),
  );
  await assert.rejects(() => access(stalePreparation), /ENOENT/);
});

async function createRemote(): Promise<string> {
  const directory = await mkdtemp(join(tmpdir(), 'upstream-remote-'));
  await execFileAsync('git', ['init', '-b', 'main'], { cwd: directory });
  await execFileAsync('git', ['config', 'user.name', 'Upstream Fixture'], { cwd: directory });
  await execFileAsync('git', ['config', 'user.email', 'fixture@example.com'], { cwd: directory });
  await write(directory, 'content/SKILL.md', '---\nname: example\ndisable-model-invocation: true\n---\nupstream\n');
  await write(directory, 'manifest/plugin.json', '{"version":"2.4.0"}');
  await execFileAsync('git', ['add', '.'], { cwd: directory });
  await execFileAsync('git', ['commit', '-m', 'fixture'], { cwd: directory });
  return directory;
}

async function write(root: string, relative: string, content: string): Promise<void> {
  const path = join(root, relative);
  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, content);
}

function configFor(repository: string, ref: string): string {
  return JSON.stringify({
    upstreams: [
      {
        name: 'example',
        repository,
        ref,
        copies: [
          {
            source: 'content',
            destination: 'plugins/example/skills',
            removeFrontmatter: ['disable-model-invocation'],
          },
        ],
        version: {
          source: 'manifest/plugin.json',
          targets: ['plugins/example/.codex-plugin/plugin.json'],
          appendCommit: true,
        },
      },
    ],
  });
}
